### PR TITLE
Add VP9 Hardware Encoder

### PR DESCRIPTION
**Pull request by onabiull**
Signed-off-by: Nabiullin, Oleg <oleg.nabiullin@intel.com>

### CODE REVIEW

File: _studio/CMakeLists.txt [hunk 1]
```diff
@@ -32,5 +32,6 @@ add_subdirectory(mfx_lib)
 if(MFX_ENABLE_USER_ENCODE)
   add_subdirectory(hevce_hw)
   add_subdirectory(hevc_fei)
+  add_subdirectory(vp9e_hw)
```
[the remark this answers is not in the document]
Comment (onabiull) [reply _studio/CMakeLists.txt]: We did release VP9e as a plugin at least as part of MediaSDK for Windows. And more importantly this fragment currently needed for our internal validation tools and samples.

File: _studio/mfx_lib/encode_hw/vp9/include/mfx_platform_defs.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef __MFX_PLATFORM_DEFS_H__
```
[the remark this answers is not in the document]
Comment (onabiull) [reply _studio/mfx_lib/encode_hw/vp9/include/mfx_platform_defs.h]: Will do

File: _studio/mfx_lib/encode_hw/vp9/include/mfx_vp9_encode_hw.h [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    {
+        return m_initialized;
+    }
+
+protected:
+    VP9MfxVideoParam              m_video;
+    std::list<VP9MfxVideoParam>   m_videoForParamChange; // encoder keeps several versions of encoding parameters
+                                                         // to allow dynamic parameter change w/o drain of all buffered tasks.
+                                                         // Tasks submitted before the change reference to previous parameters version.
+    std::auto_ptr <DriverEncoder> m_ddi;
```
[the remark this answers is not in the document]
Comment (onabiull) [reply _studio/mfx_lib/encode_hw/vp9/include/mfx_vp9_encode_hw.h]: ok

File: _studio/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (stasefremov) [reply _studio/CMakeLists.txt]: AFAIK we haven't ever released VP9 encoder yet.

File: _studio/mfx_lib/encode_hw/vp9/include/mfx_vp9_encode_hw_utils.h [hunk 4]
```diff
@@ -0,0 +1,1020 @@
...
+#ifdef VP9_LOGGING
+#define VP9_LOG(string, ...) printf(string, ##__VA_ARGS__); fflush(0);
+#else
+#define VP9_LOG(string, ...)
+#endif
+
+namespace MfxHwVP9Encode
+{
+
+#define ALIGN(x, align) (((x) + (align) - 1) & (~((align) - 1)))
```
[the remark this answers is not in the document]
Comment (onabiull) [reply _studio/mfx_lib/encode_hw/vp9/include/mfx_vp9_encode_hw_utils.h]: ok

File: _studio/mfx_lib/encode_hw/vp9/include/mfx_vp9_encode_hw_utils.h [hunk 5]
```diff
@@ -0,0 +1,1020 @@
...
+#define ALIGN_POWER_OF_TWO(value, n) \
+    (((value)+((1 << (n)) - 1)) & ~((1 << (n)) - 1))
+#define MFX_CHECK_WITH_ASSERT(EXPR, ERR) { assert(EXPR); MFX_CHECK(EXPR, ERR); }
+
+#define DPB_SIZE 8 // DPB size by VP9 spec
```
[the remark this answers is not in the document]
Comment (onabiull) [reply _studio/mfx_lib/encode_hw/vp9/include/mfx_vp9_encode_hw_utils.h]: I suggest to implement most of non-trivial refactorings as subsequent commits.

File: _studio/mfx_lib/encode_hw/vp9/include/mfx_vp9_encode_hw_utils.h [hunk 6]
```diff
@@ -0,0 +1,1020 @@
...
+                mfxStatus sts = FreeSurface(pFrame, pCore);
+                MFX_CHECK_STS(sts);
+            }
+        }
+
+        return MFX_ERR_NONE;
+    }
+
+#define MFX_CHECK_WITH_ASSERT(EXPR, ERR) { assert(EXPR); MFX_CHECK(EXPR, ERR); }
+#define STATIC_ASSERT(ASSERTION, MESSAGE) char MESSAGE[(ASSERTION) ? 1 : -1]; MESSAGE
```
[the remark this answers is not in the document]
Comment (onabiull) [reply _studio/mfx_lib/encode_hw/vp9/include/mfx_vp9_encode_hw_utils.h]: ok

**Pull request merged**